src/shims/unix/fs.rs

```diff
@@ -1070,24 +1072,41 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             Some(Ok(dir_entry)) => {
                 // Write the directory entry into a newly allocated buffer.
                 // The name is written with write_bytes, while the rest of the
-                // dirent64 struct is written using write_int_fields.
+                // dirent64 (or dirent) struct is written using write_int_fields.
 
                 // For reference:
+                // On Linux:
                 // pub struct dirent64 {
                 //     pub d_ino: ino64_t,
                 //     pub d_off: off64_t,
                 //     pub d_reclen: c_ushort,
                 //     pub d_type: c_uchar,
                 //     pub d_name: [c_char; 256],
                 // }
+                //
+                // On Solaris:
+                // pub struct dirent {
+                //     pub d_ino: ino64_t,
+                //     pub d_off: off64_t,
+                //     pub d_reclen: c_ushort,
+                //     pub d_name: [c_char; 3],
+                // }
 
                 let mut name = dir_entry.file_name(); // not a Path as there are no separators!
                 name.push("\0"); // Add a NUL terminator
                 let name_bytes = name.as_encoded_bytes();
                 let name_len = u64::try_from(name_bytes.len()).unwrap();
 
-                let dirent64_layout = this.libc_ty_layout("dirent64");
-                let d_name_offset = dirent64_layout.fields.offset(4 /* d_name */).bytes();
+                let is_linux = matches!(&*this.tcx.sess.target.os, "linux");
+
+                let dirent64_layout = if is_linux {
```

Comment:
My suggestion was to pass the type name in as a string. IMO that makes more sense since both dirent64 and dirent exist on Linux.

Why did you deviate from my suggestion? Please don't just silently do something different and ask for the next review. That is tiring as I have to painstakingly check for things like that. Instead, write a comment explaining why you deviated from the proposed plan.

Reply:
No I just did not get you (or forgot to address it). I ll look at it shortly.